merry-go-round.md

```diff
@@ -0,0 +1,133 @@
+  be able to *update* the data they have already received to newer and newer
+  state roots.
+
+To address this, we will introduce two changes to how nodes in the network
+handle state.
+
+First, we introduce a concept of epoch boundaries.  Nodes in the network will
+store proof of the segment of the state that they are storing for the network
+at the most recent state root from the most recent epoch boundary.  The result
+of this is that within a single epoch, all requests for state data from
+different nodes will be anchored to the same state root.
+
+Under this model we still assume that all updates to the state are being
+gossiped as each block is added to the chain.  Nodes that are storing data
+would accumulate a state diff from these proofs during the course of an epoch,
+and at the end of the epoch, they can apply this state diff to their stored
+state data to update their stored state to the new state root at the new epoch
+boundary.
+
+Research should be done to determine optimal epoch lengths.  The epoch length
+must be short enough that a node can fully traverse the full DHT within a
+single epoch.  We suggest 256 (51 minutes) as a starting point.
+
+## Specification
+
+### Storage
+
+Nodes in the network store two things.
+
+* proofs about the state *near* their position in the DHT anchored at the most
+  recent epoch boundary
+* diffs about the state accumulated from gossip that allow updating the state
+  from the most recent epoch boundary to the newest state root.
+
+> Note, that nodes will likely need to store at least the most recent two
+> epochs of state data to account for the points in time where we are
+> transitioning between epochs.  Nodes should continue to be able to retrieve
+> data from the previous epoch to allow them the necessary time to fully obtain
+> the necessary proof data to update their view of the state to the newest
+> epoch.
+
+### Gossip
+
+The network provides a gossip mechanism that disseminates new state data from
+each block to the nodes in the network that would store that data.
+
+### Retrieval
+
+The network exposes request/response primatives for:
+
+* retrieving a node's proof of the state at a recent epoch boundary, likely bounded by a range of keys within the trie.
+* retrieving a a node's proof diff that allows you to update their state proof to either a newer state root or the latest epoch boundary.
+
+
+### Syncing
+
+The process for syncing then becomes the following.
+
+1. Let `P` be our proof against the state at the most recent epoch boundary.  The initial value of `P` is an empty proof.
+2. Let `L` be the lowest key within the trie that we are missing leaf data for.  The initial value of `L` would be the `0x0` key.
+3. Repeat the following until the head of the chain passes an epoch boundary.
+    A. Find a node that is close to `L` and request it's proof data which will be anchored to the most recent epoch boundary.
```

Comment:
Whenever you mention proof data, do you mean a range of leaves + their boundary proofs against the state root (of that recent epoch boundary)?

Reply:
correct.  (work in progress to find clear terminology for this stuff)